internal/controller/managedcluster_controller.go

```diff
@@ -15,10 +15,12 @@
 package controller
```

Comment:
general suggestion and a couple of nit comments:

* I think we can separate the new code into a new package, IMO it'd be easier to manage (it does not matter whether it would be implemented on interfaces or just public fixed structs)
* `kube-system` consts and etc. -> `metav1.NamespaceSystem`
* complicated list options might be simplified with `"sigs.k8s.io/controller-runtime/pkg/client".InNamespace`, `"sigs.k8s.io/controller-runtime/pkg/client".HasLabels`, `"sigs.k8s.io/controller-runtime/pkg/client".Limit` and etc.
* some errors might be wrapped (`%w`) where it might make more sense than a plain string (`%s`)
* some variables (such as secret names) might be instantiated as constants
* some error-typed return arguments might be annotated/wrapped with an understandable message (`return err` -> `return fmt.Errorf("failed to perform some action during another event: %w", err)` to ease potential debugging 
* I think some of the configurations (e.g. azure/vsphere/etc.) might be commented on for future readers, now it can be only understandable by users with an in-depth experience, i.e. I know why `insecureFlag` is required for the vSphere CSI but one cannot
* personal comment (so skip it if it's not okay for you): I find reading code with `\n` between logical blocks much more convenient than reading uninterrupted huge blocks of code

Reply:
I will try to address these in more or less the same order:
1. I don't see benefits of a separate package: it makes some methods and data a bit more complicated and benefit is near zero. Probably we can move new code to a separate file, but leave in the same package.
2. Done
3. Chain method calls looks more complicated that passing struct as an argument. But I guess it's more like a personal preference. I will leave it as is if you not insist on that.
4. Please provide exact places where it will make more sense
5. If you're about CCM secret names I don't see why they should be in global scope. I specifically reduced their scope to one function.
6. Please give me specific example of such place
7. It will be a lot to comment (I don't like this idea). Maybe it's better to document it somewhere in the actual docs?
8. I didn't think about it and tbh I don't even see what was bothering you

Reply:
1. The following points actually:
* compilation point of view: having a gigantic package slows the compile time
* LSP point of view: having a gigantic package with lots of symbols slows the initiation and the following search
* code/developer point of view: clean code with separated responsibilities; symbols like `internal/controller.makeClientFromSecret` or `internal/controller.applyCCMConfigs` (and others that have been introduced) are used purely within the managed-cluster-ctrl, and a developer may misuse them (we have almost a dozen of controllers in the package) or spend time to read the code of the symbols whether those might be reused or it is better to implement another function/method at the moment

This point is not mandatory to implement within this PR and could be implemented as part of a follow-up PR. I also might be wrong with the aforementioned points thus we can ask for another opinion on that point

3. Right, that's mostly personal preferences, for me personally the 
```go
&client.ListOptions{
 // ...
  LabelSelector: labels.SelectorFromSet(map[string]string{
    hmc.ClusterNameLabelKey: managedCluster.Name,
  }),
  // ...
},
```
looks not just as passing a struct compared to the following:
```go
client.HasLabels{hmc.ClusterNameLabelKey: managedCluster.Name}
```
4. Sure, for example, the 
```go
fmt.Errorf("failed to get cluster providers for cluster %s/%s: %s" /* ... */
```
converts the underlying error (which might be 4xx/5xx) to a plain string and a caller won't be able to extract any details from it if it wanted to. This is mostly speculation, but if that happens and a caller wants to deconstruct the returned error, a developer should then change other pieces of code from `%s` to the `%w`. There are different places where an error converts to a plain string, most of them — API calls (`get`/`list`)
5. I did not mention the **global** scope, no one stops from converting `secretName := "vsphere-cloud-secret"` to the e.g. `const vsphereCloudSecretName = "vsphere-cloud-secret"` within the same line (applicable elsewhere **if any**)
6. That bullet point mostly relates to the `makeClientFromSecret` function, each error could be wrapped with a detailed message but you may skip this if `failed to create k8s client: <underlying message>` is more suitable for you than e.g. `failed to create k8s client: failed to get REST config from the <name> secret: <underlying message>`.
7. Sounds good and useful, especially for potential readers of the docs, they will know the defaults and why we have such defaults 
8. Nvm, I see now more new lines between logical blocks have been added, it's okay now

Reply:
Agreed with @Kshatrix upon a follow-up PR

Reply:
Separate task on that #530
@zerospiel If you want to add something there feel free to do so. Thanks